c7n/resources/ec2.py

```diff
@@ -918,6 +918,67 @@ def process(self, instances):
                 Groups=groups[idx])
 
 
+@actions.register('set-instance-profile')
+class SetInstanceProfile(BaseAction):
+    """Sets (or removes) the instance profile for an existing EC2 instance.
+
+    :Example:
+
+    .. code-block: yaml
+
+        policies:
+          - name:
+            resource: ec2
+            query:
+              - IamInstanceProfile: absent
+            actions:
+              - type: set-instance-profile
+                name: default
+
+    https://docs.aws.amazon.com/cli/latest/reference/ec2/associate-iam-instance-profile.html
+    https://docs.aws.amazon.com/cli/latest/reference/ec2/disassociate-iam-instance-profile.html
+    """
+
+    schema = type_schema(
+        'set-instance-profile',
+        **{'name': {'type': 'string'}})
+
+    permissions = (
+        'ec2:AssociateIamInstanceProfile',
+        'ec2:DisassociateIamInstanceProfile',
+        'iam:PassRole')
+
+    def process(self, instances):
+        client = utils.local_session(
+            self.manager.session_factory).client('ec2')
+        profile_name = self.data.get('name', '')
+
+        for i in instances:
+            if profile_name:
+                client.associate_iam_instance_profile(
+                    IamInstanceProfile={'Name': self.data.get('name', '')},
+                    InstanceId=i['InstanceId'])
+            else:
+                response = client.describe_iam_instance_profile_associations(
```

Comment:
instance profile is already available as an attribute of the instance per describe output, as IamInstanceProfile